cpp/bench/sg/dbscan.cu

```diff
@@ -0,0 +1,116 @@
+/*
+ * Copyright (c) 2019, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cuML.hpp>
+#include <dbscan/dbscan.hpp>
+#include <utility>
+#include "benchmark.cuh"
+
+namespace ML {
+namespace Bench {
+namespace dbscan {
+
+struct AlgoParams {
+  int min_pts;
+  double eps;
+  size_t max_bytes_per_batch;
+};
+
+struct Params {
+  DatasetParams data;
+  BlobsParams blobs;
+  AlgoParams dbscan;
+};
+
+template <typename D>
+class Dbscan : public BlobsFixture<D> {
+ public:
+  Dbscan(const std::string& name, const Params& p)
+    : BlobsFixture<D>(p.data, p.blobs), dParams(p.dbscan) {
+    this->SetName(name.c_str());
+  }
+
+ protected:
+  void runBenchmark(::benchmark::State& state) override {
+    if (!this->params.rowMajor) {
+      state.SkipWithError("Dbscan only supports row-major inputs");
+    }
+    auto& handle = *this->handle;
+    auto stream = handle.getStream();
+    for (auto _ : state) {
+      CudaEventTimer timer(handle, state, true, stream);
+      dbscanFit(handle, this->data.X, this->params.nrows, this->params.ncols,
+                D(dParams.eps), dParams.min_pts, labels,
+                dParams.max_bytes_per_batch);
+    }
+  }
+
+  void allocateBuffers(const ::benchmark::State& state) override {
+    auto allocator = this->handle->getDeviceAllocator();
+    auto stream = this->handle->getStream();
+    labels =
+      (int*)allocator->allocate(this->params.nrows * sizeof(int), stream);
+  }
+
+  void deallocateBuffers(const ::benchmark::State& state) override {
+    auto allocator = this->handle->getDeviceAllocator();
+    auto stream = this->handle->getStream();
+    allocator->deallocate(labels, this->params.nrows * sizeof(int), stream);
+  }
+
+ private:
+  AlgoParams dParams;
+  int* labels;
+};
+
+std::vector<Params> getInputs() {
+  std::vector<Params> out;
+  Params p;
+  p.data.rowMajor = true;
+  p.blobs.cluster_std = 1.0;
+  p.blobs.shuffle = false;
+  p.blobs.center_box_min = -10.0;
+  p.blobs.center_box_max = 10.0;
+  p.blobs.seed = 12345ULL;
+  p.dbscan.max_bytes_per_batch = 0;
+  std::vector<std::pair<int, int>> rowcols = {
+    {10000, 81},
+    {20000, 128},
+    {40000, 128},
```

Comment:
We should probably be benchmarking on the orders of 1k, 5k, 10k, 50k, 100k, 1M

Reply:
fair points.
1M sounds a bit large in terms of overall runtime, no? Can we, maybe, limit ourselves to 100k?

Reply:
100k is fine. I recently started pushing to see how dbscan does on >1M because Joe Eaton recently brought MuDBSCAN to my attention, which claims to cluster 1B samples on 32 nodes in 41mins. I looked at their GitHub and noticed their code is already written in MPI. We might want to consider this as a potential starting point for our MNMG DBSCAN. Tagging @akkamesh, as we might be working on it together.